Доработана функциональность по ТЗ спринта 6

iGladkikh opened the pull request.

iGladkikh committed:
Изменить логину работы истории:
1. Сделать историю посещений неограниченной по размеру.
2. Избавиться от повторных просмотров в истории.

iGladkikh committed:
Защитить данные внутри менеджера от изменений при вызове сеттеров объектов.

Yang-Pi commented on src/services/InMemoryHistoryManager.java:
```diff
@@ -1,27 +1,33 @@
 
     public InMemoryHistoryManager() {
-        this.history = new ArrayList<>(HISTORY_MAX_ITEMS_COUNT);
+        this.history = new LinkedHashMap<>();
```
Все верно, только класс связного списка должен быть не из библиотеки java.util, а самописный)

iGladkikh replied:
Исправлено. Спасибо!

Yang-Pi commented on src/services/InMemoryTaskManager.java:
```diff
@@ -48,23 +51,23 @@ private void calculateEpicStatus(Epic epic) {
 
         int epicId = epic.getId();
-        epics.put(epicId, epic);
+        epics.put(epicId, epic.copy());
```
Лишняя операция на самом деле, понимаю твою идею, но сам объект - это фактически модель данных, полученная по API (далее вы именно это и будете делать)

iGladkikh replied:
Исправлено. Спасибо!

iGladkikh committed:
Внести исправление после ревью 6.1:
1. Класс связного списка истории должен быть не из библиотеки java.util.
2. Исключить метод "copy" при добавлении/получении объектов задач.

Yang-Pi commented on src/services/InMemoryHistoryManager.java:
```diff
@@ -1,27 +1,105 @@
+
+        ArrayList<Task> result = new ArrayList<>(history.size());
+        Node node = last;
+        result.add(node.data);
```
node вполне мб null, поэтому получишь NPE тут

iGladkikh replied:
Исправлено. Дополнена проверка в стр. 44. Спасибо!

Yang-Pi commented on src/services/InMemoryHistoryManager.java:
```diff
@@ -1,27 +1,105 @@
+            result.add(node.prev.data);
+            node = node.prev;
+        }
+        return result;
```
По ТЗ список задач нужно вернуть в том порядке, как они добавлялись, у тебя же он обратный

iGladkikh replied:
Исправлено. Спасибо!

Yang-Pi commented on src/services/InMemoryHistoryManager.java:
```diff
@@ -1,27 +1,105 @@
 public class InMemoryHistoryManager implements HistoryManager {
-    private static final int HISTORY_MAX_ITEMS_COUNT = 10;
-    private final List<Task> history;
+    private Node last;
```
Рекомендую все же добавить еще и голову списка (head & tail или first & last)

iGladkikh replied:
Добавлено поле Node first. Спасибо!

iGladkikh committed:
Внести исправление после ревью 6.2:
1. Добавить переменную first в класс InMemoryHistoryManager.
2. Список задач в истории нужно вернуть в том порядке, как они добавлялись.
3. Проверить node на null при добавлении в историю.

Yang-Pi commented on src/services/InMemoryHistoryManager.java:
```diff
@@ -1,27 +1,120 @@
+        ArrayList<Task> result = new ArrayList<>(history.size());
+        Node node = first;
+        result.add(node.data);
+        while (node.hasNext()) {
```
Работает, но сложно)
```
Node tmp = first;
while (tmp != null) {
...
tmp = tmp.next();
}
```

iGladkikh replied:
Исправлено.

iGladkikh committed:
Внести исправление после ревью 6.3:
1. Оптимизировать метод getHistory класса InMemoryHistoryManager.
